### PR TITLE
Rename `dealloc_ref [stack]` and enable StackPromotion for OSSA

### PR DESCRIPTION
**Pull request by eeckstein**
The blocker for enabling StackPromotion for OSSA was that the ownership of `dealloc_ref [stack]` was wrong. It should have `NonUse` ownership, because it's just an end-of-lifetime marker for an `alloc_ref [stack]`.

Also, the instruction syntax `dealloc_ref [stack]` is very confusing, because it does not do anything like a deallocation.
It's renamed to `dealloc_stack_ref `, which better describes its purpose.

Renaming means: removed the `[stack]` flag from `dealloc_ref` and introduced a new instruction `dealloc_stack_ref ` (with correct operand ownership of `NonUse`)

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

File: docs/SIL.rst [hunk 1]
```diff
@@ -3381,13 +3381,26 @@ project_box
 
 Given a ``@box T`` reference, produces the address of the value inside the box.
 
+end_stack_ref_lifetime
+``````````````````````
+::
+
+  sil-instruction ::= 'dealloc_ref' sil-operand
+
+  dealloc_ref %0 : $T
```
Comment (benrimmington) [inline docs/SIL.rst:3390]:
```suggestion
  sil-instruction ::= 'end_stack_ref_lifetime' sil-operand

  end_stack_ref_lifetime %0 : $T
```

Comment (eeckstein) [reply docs/SIL.rst:3390]: thanks!

**Comment by atrick** [comment]
Meghana fixed this here, then closed the PR:
https://github.com/apple/swift/pull/39803

`dealloc_ref [stack]` semantically deallocates stack space for a deinitialized object (of course there's no runtime deallocation for stack deallocation). `dealloc_ref [stack]` has has nothing to do with the object's lifetime. Except that the lifetime must obviously end before storage is deallocated.

If we care about the SIL making sense, then it would look like this:

```
%ref, %addr = alloc_stack_ref $Class // %ref and %addr are physically the same pointers
…
consume %ref     // in practice, this is always an @owned apply argument
dealloc_stack %addr : $Class
```

After devirtualization:

```
%ref, %addr = alloc_stack_ref $Class // %ref and %addr are physically the same pointers
…
set_deallocating %ref
end_lifetime %ref     // nop disappears after lowering ownership
dealloc_stack %addr : $Class
```

The OSSA lifetime utilities might end up working in practice if `dealloc_stack` takes the reference value as a `NonUse` and we're careful to check NonUse everywhere instead of isTypeDependentOperand. But there may also be places we iterate over the non-typedependent operands and assume they are real uses (which I think is a nice invariant). Again, those places could be fixed case-by-case, but that's a downside to having an illogical hack in the representation. That just needs to be weighed against the extra work involved in implementing a representation that does make sense. The PR author can decide that.

Regardless, `end_stack_ref_lifetime` is the wrong name because the object lifetime must end before this instruction. The instruction just marks the deallocation of stack space.

**Comment by eeckstein** [comment]
I though about the two-return-value variant of alloc_ref [stack], too. But this would be a massive change and would complicate many other parts of the SIL and optimizer. IMO we don't have to over-engineer this.

@atrick what do you think of `dealloc_stack_ref`? (I don't want to use the existing `dealloc_stack` because that's assumed to be paired to `alloc_stack` in so many places in the compiler)

**Comment by atrick** [comment]
> I though about the two-return-value variant of alloc_ref [stack], too. But this would be a massive change and would complicate many other parts of the SIL and optimizer. IMO we don't have to over-engineer this.

That's a reasonable position. We shouldn't block forward progress on doing that work.

> @atrick what do you think of `dealloc_stack_ref`? (I don't want to use the existing `dealloc_stack` because that's assumed to be paired to `alloc_stack` in so many places in the compiler)

Yes, that's fine!

I'm also fine with just adding a property to dealloc_stack that can be checked by the code that searches for `alloc_stack`. I can't imagine it's many places. And I wonder if it makes sense to be looking for `alloc_ref [stack]` in those cases. 

Then we'd have

(1) `dealloc_stack`
Where a non-address argument type implies that the stack location was allocated and initialized together

(2) `dealloc_stack [ref]`

Has exactly same issue as #1. Just a bit more explicit

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Review by atrick — APPROVED** [review]

**Pull request merged**